### PR TITLE
feat: switch to poetry dependency management

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

This PR switches from `pipenv` to `poetry` for dependency and virtual environment management.

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](https://github.com/fetchai/agents-aea/blob/main/CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -196,39 +188,6 @@ jobs:
...
-          pip install tox
-      - name: Check Pipfile and tox.ini consistency
-        run : |
-          python ./scripts/check_pipfile_and_toxini.py
-      - name: Install current aea
-        run: |
-          pip install -e .[all]
-      - name: Check aea dependenices and imports
-        run: |
-          tox -e dependencies_check
```
Comment (5A11) [inline .github/workflows/workflow.yml:230]: Think we can continue using this. Just have to adapt it to pyproject.toml

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go [hunk 2]
```diff
@@ -1937,7 +1936,7 @@ func expectEnvelopeOrdered(t *testing.T, rx chan *aea.Envelope, counter int) {
 		}
 		message, _ := strconv.Atoi(string(envel.Message))
 		if message != counter {
-			log.Fatal(fmt.Sprintf("Expected counter %d received counter %d", counter, message))
+			log.Fatalf("Expected counter %d received counter %d", counter, message)
 		}
 	case <-timeout:
 		t.Error("Failed to receive envelope before timeout")
```
Comment (5A11) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:1942]: Was this missed in the previous Go Lang PR?

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (5A11) [inline .github/workflows/workflow.yml:230]: I think wee can still use this script. Just need to adapt it to pyproject.toml.

File: .github/workflows/workflow.yml [hunk 3]
```diff
@@ -299,7 +258,6 @@ jobs:
     - common_checks_2
     - common_checks_3
     - common_checks_4
-    - dependencies_checks
```
Comment (5A11) [inline .github/workflows/workflow.yml:302]: Same as above.

File: .github/workflows/workflow.yml [hunk 4]
```diff
@@ -346,7 +304,6 @@ jobs:
       - common_checks_2
       - common_checks_3
       - common_checks_4
-      - dependencies_checks
```
Comment (5A11) [inline .github/workflows/workflow.yml:349]: Same as above.

File: .github/workflows/workflow.yml [hunk 5]
```diff
@@ -382,13 +339,12 @@ jobs:
     - common_checks_2
     - common_checks_3
     - common_checks_4
-    - dependencies_checks
```
Comment (5A11) [inline .github/workflows/workflow.yml:385]: Same as above.

File: .github/workflows/workflow.yml [hunk 6]
```diff
@@ -461,7 +417,6 @@ jobs:
...
```
Comment (5A11) [inline .github/workflows/workflow.yml:464]: Same as above.

File: packages/fetchai/skills/tac_control/helpers.py [hunk 7]
```diff
@@ -313,9 +313,9 @@ def generate_equilibrium_prices_and_holdings(  # pylint: disable=unused-argument
...
     # maths
-    endowments_a = np.array(good_endowments_l, dtype=np.int)
-    scaled_utility_params_a = np.array(
-        utility_params_l, dtype=np.float
+    endowments_a = np.array(good_endowments_l, dtype=np.int)  # type: ignore
+    scaled_utility_params_a = np.array(  # type: ignore
+        utility_params_l, dtype=np.float  # type: ignore
     )  # note, they are already scaled
     endowments_by_good = np.sum(endowments_a, axis=0)
     scaled_params_by_good = np.sum(scaled_utility_params_a, axis=0)
```
Comment (5A11) [inline packages/fetchai/skills/tac_control/helpers.py:321]: Take out. Address in mypy PR.

File: packages/fetchai/skills/tac_control_contract/helpers.py [hunk 8]
```diff
@@ -287,9 +287,9 @@ def generate_equilibrium_prices_and_holdings(  # pylint: disable=unused-argument
...
     # maths
-    endowments_a = np.array(good_endowments_l, dtype=np.int)
-    scaled_utility_params_a = np.array(
-        utility_params_l, dtype=np.float
+    endowments_a = np.array(good_endowments_l, dtype=np.int)  # type: ignore
+    scaled_utility_params_a = np.array(  # type: ignore
+        utility_params_l, dtype=np.float  # type: ignore
     )  # note, they are already scaled
     endowments_by_good = np.sum(endowments_a, axis=0)
     scaled_params_by_good = np.sum(scaled_utility_params_a, axis=0)
```
Comment (5A11) [inline packages/fetchai/skills/tac_control_contract/helpers.py:295]: Take out and address in mypy PR.

File: .github/workflows/workflow.yml [hunk 9]
```diff
@@ -16,22 +16,14 @@ jobs:
...
           sudo apt-get autoclean
           sudo pip install pipenv
-      - name: Pipenv install requirements and check it can be locked
-        run: |
-          pipenv install --dev --skip-lock --clear
-          pipenv run pip install -e .[all]
-          pipenv run pip install --no-deps file:plugins/aea-ledger-ethereum
-          pipenv run pip install --no-deps file:plugins/aea-ledger-cosmos
-          pipenv run pip install --no-deps file:plugins/aea-ledger-fetchai
-          pipenv lock
```
[the remark this answers is not in the document]
Comment (5A11) [reply .github/workflows/workflow.yml:34]: Lock is already there. poetry will complain if lock is out of sync with deps in pyproject.toml

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (5A11) [reply .github/workflows/workflow.yml:230]: This script is no compatible with poetry (read setup.py). Need to adapt it to poetry (in another PR) and enable it.

File: Makefile [hunk 10]
```diff
@@ -106,16 +106,12 @@ test-all:
 
 .PHONY: install
 install: clean
-	python3 setup.py install
+	ppip install .[all]
```
Comment (5A11) [inline Makefile:109]: typo

File: pyproject.toml [hunk 11]
```diff
@@ -1,3 +1,130 @@
...
+name = "aea"
+version = "1.2.1"
+description = "Autonomous Economic Agent framework"
+authors = ["Fetch.AI Limited"]
+readme = "README.md"
+license = "Apache-2.0"
+homepage = "https://github.com/fetchai/agents-aea"
+classifiers = [
+    "Environment :: Console",
+    "Environment :: Web Environment",
```
Comment (5A11) [inline pyproject.toml:16]:
```suggestion
```

File: pyproject.toml [hunk 12]
```diff
@@ -1,3 +1,130 @@
...
+    "Programming Language :: Python :: 3.10",
+    "Topic :: Communications",
+    "Topic :: Internet",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Software Development",
+    "Topic :: System",
+]
+
+[tool.poetry.scripts]
+aea = 'aea.cli:cli'
```
Comment (5A11) [inline pyproject.toml:35]: Why in scripts?

File: pyproject.toml [hunk 13]
```diff
@@ -1,3 +1,130 @@
...
+pywin32 = { version = "==303", markers = "sys_platform == 'win32' or platform_system == 'Windows'" }
+
+# cli, test_tools extras
+click = { version = ">=7.0.0,<8.0.0", optional = true }
+
+
+[tool.poetry.extras]
+cli = ["click"]
+test_tools = ["click"]
+all=["click"]
```
Comment (5A11) [inline pyproject.toml:59]: Do we need test_tools and all extras?

File: pyproject.toml [hunk 14]
```diff
@@ -1,3 +1,130 @@
...
+
+
+[tool.poetry.group.tools.dependencies]
+#benchmark
+matplotlib = ">=3.3.0,<3.4"
+
+gitpython = ">=3.1.14"
+ipfshttpclient = "==0.6.1"
+memory-profiler = "==0.57.0"
+psutil = "==5.7.0"
```
Comment (5A11) [inline pyproject.toml:130]: Why these grouping generally?

Comment (5A11) [reply pyproject.toml:130]: Also there are some duplicates in the groups. Is that alright?

File: tox.ini [hunk 15]
```diff
@@ -7,127 +7,58 @@
...
-    ; for password encryption in cosmos
-    pycryptodome==3.15.0
-    blspy==1.0.14
-	bip-utils==2.5.1
-	python-mbedtls==2.2.0
-	py-sr25519-bindings==0.1.4
-    cosmpy>=0.5.0,<0.6.0
-
-
+deps = toml
```
Comment (5A11) [inline tox.ini:19]: why is this needed?

File: tox.ini [hunk 16]
```diff
@@ -7,127 +7,58 @@
...
-
-[testenv:py3.6]
-basepython = python3.6
-deps =
-    {[testenv]deps}
-    tensorflow==2.4.0
+    poetry install --only testing --only main --only packages -E all
+    poetry run pip install openapi-core==0.13.2
+   	poetry run pip install {toxinidir}[all]
+    poetry run python ./install_packages.py jsonschema
```
Comment (5A11) [inline tox.ini:24]: Why manually add `openapi-core` and `jsonschema`?

File: tox.ini [hunk 17]
```diff
@@ -7,127 +7,58 @@
...
+    poetry run python -m pip install file://{toxinidir}/plugins/aea-ledger-ethereum
+    poetry run python -m pip install file://{toxinidir}/plugins/aea-ledger-cosmos
+    poetry run python -m pip install file://{toxinidir}/plugins/aea-ledger-fetchai
+    poetry run python -m pip install file://{toxinidir}/plugins/aea-cli-ipfs
+    poetry run python -m pip install typed-ast -U
```
Comment (5A11) [inline tox.ini:29]: Same as above.

File: tox.ini [hunk 18]
```diff
@@ -141,58 +72,63 @@ commands = {[plugins]commands}
...
+    poetry run bandit -s B101 -r tests scripts
 
 [testenv:black]
 skipsdist = True
 skip_install = True
-deps =
-    black==19.10b0
-commands = black aea benchmark examples packages plugins scripts tests
+commands = 
+  poetry run python ./install_packages.py click black
```
Comment (5A11) [inline tox.ini:96]: Need `click`?

Comment (5A11) [reply tox.ini:96]: Also below.

File: tox.ini [hunk 19]
```diff
@@ -209,103 +145,74 @@ commands = {toxinidir}/scripts/check_packages.py
...
-    packaging>=20.4,<=21.5
-    web3==5.12.0
-    yoti==2.14.0
-    asn1crypto==1.4.0
-commands = mypy aea packages --disallow-untyped-defs
-           mypy benchmark examples  --check-untyped-defs
-           mypy scripts tests
+commands = 
+    poetry install --only packages,main,testing
+    poetry run python ./install_packages.py mypy click
```
Comment (5A11) [inline tox.ini:180]: click?

File: tox.ini [hunk 20]
```diff
@@ -209,103 +145,74 @@ commands = {toxinidir}/scripts/check_packages.py
...
-    pytest==7.0.0
-    gitpython>=3.1.14
-
 commands =
-    python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-ethereum
-    python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-cosmos
-    python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-fetchai
-    python -m pip install --no-deps file://{toxinidir}/plugins/aea-cli-ipfs
-    sh -c "pylint aea benchmark packages plugins/aea-ledger-fetchai/aea_ledger_fetchai plugins/aea-ledger-ethereum/aea_ledger_ethereum plugins/aea-ledger-cosmos/aea_ledger_cosmos plugins/aea-cli-ipfs/aea_cli_ipfs scripts examples/*"
+    poetry run python ./install_packages.py pylint pytest gitpython
```
Comment (5A11) [inline tox.ini:189]: why pytest and gitpython?

File: tox.ini [hunk 21]
```diff
@@ -209,103 +145,74 @@ commands = {toxinidir}/scripts/check_packages.py
...
+    poetry run python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-ethereum
+    poetry run python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-cosmos
+    poetry run python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-fetchai
+    poetry run python -m pip install --no-deps file://{toxinidir}/plugins/aea-cli-ipfs
+    poetry run sh -c "pylint aea benchmark packages plugins/aea-ledger-fetchai/aea_ledger_fetchai plugins/aea-ledger-ethereum/aea_ledger_ethereum plugins/aea-ledger-cosmos/aea_ledger_cosmos plugins/aea-cli-ipfs/aea_cli_ipfs scripts examples/*"
```
Comment (5A11) [inline tox.ini:194]: why is this different to other commands (has to use shell)?

File: tox.ini [hunk 22]
```diff
@@ -7,127 +7,58 @@
...
-    python -m pip install --no-deps file://{toxinidir}/plugins/aea-ledger-fetchai
-    python -m pip install --no-deps file://{toxinidir}/plugins/aea-cli-ipfs
-    pytest -rfE tests/ --cov=aea --cov=packages/fetchai/connections --cov=packages/fetchai/contracts --cov=packages/fetchai/protocols --cov=packages/fetchai/skills --cov-report=html --cov-report=xml --cov-report=term --cov-report=term-missing --cov-config=.coveragerc {posargs}
-
-[testenv:py3.6]
-basepython = python3.6
-deps =
-    {[testenv]deps}
-    tensorflow==2.4.0
+    poetry install --only testing --only main --only packages -E all
```
Comment (5A11) [inline tox.ini:21]: use comma

File: pyproject.toml (hunk 12 above)
Comment (5A11) [reply pyproject.toml:35]: List of executable, in this case the aea cli.

File: pyproject.toml (hunk 13 above)
Comment (5A11) [reply pyproject.toml:59]: Try to remove cli and test_tools if they are not explicitly used. They are redundant to all. 

We'll have:

- `pip install aea`: to get aea + all basic deps minus click
- `pip install aea[all]`: to get aea + all basic deps + click

File: pyproject.toml (hunk 14 above)
Comment (5A11) [reply pyproject.toml:130]: Let's monitor this separation of deps and decide later on the suitability of this.

File: pyproject.toml [hunk 23]
```diff
@@ -1,3 +1,130 @@
...
+pydoc-markdown = "*"
+pydocstyle = "==3.0.0"
+pymdown-extensions = "==9.1"
+pygments = "==2.7.4"
+
+
+[tool.poetry.group.packages.dependencies]
+yoti = "==2.14.0"
+tensorflow = [{ version = "^2.9.2", python = "<3.11" }]
+#openapi-core = "==0.13.2"  # problematic package for poetry
```
[the remark this answers is not in the document]
Comment (5A11) [reply pyproject.toml:110]: Maybe try a newer version?

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go (hunk 2 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go]: no, it was included

File: Makefile (hunk 10 above)
Comment (solarw) [reply Makefile:109]: done

File: pyproject.toml (hunk 13 above)
Comment (solarw) [reply pyproject.toml:59]: cli will stay till we update `docs/cli-how-to.md`

test_tools - removed

File: tox.ini (hunk 15 above)
Comment (solarw) [reply tox.ini:19]: changed to poetry

File: tox.ini (hunk 18 above)
Comment (solarw) [reply tox.ini:96]: click is required to set up a specific version compatible with current black version 19.10b0

File: tox.ini (hunk 20 above)
Comment (solarw) [reply tox.ini:189]: pytest gitpython installed for scripts and aea testing tools source code check

File: tox.ini (hunk 21 above)
Comment (solarw) [reply tox.ini:194]: "examples/*" to expand exmpales automatically

File: tox.ini (hunk 22 above)
Comment (solarw) [reply tox.ini:21]: done

File: pyproject.toml [hunk 24]
```diff
@@ -1,3 +1,130 @@
...
+safety = "==1.10.3"
+mypy = "==0.761"
+darglint = "==1.8.1"
+flake8 = "==3.7.9"
+flake8-bugbear = "==20.1.4"
+flake8-docstrings = "==1.6.0"
+flake8-eradicate = "==0.4.0"
+flake8-isort = "==4.0.0"
+liccheck = "==0.6.0"
+black = "==19.10b0"
```
[the remark this answers is not in the document]
Comment (solarw) [reply pyproject.toml:82]: done

File: pyproject.toml [hunk 25]
```diff
@@ -1,3 +1,130 @@
...
+click = { version = ">=7.0.0,<8.0.0", optional = true }
+
+
+[tool.poetry.extras]
+cli = ["click"]
+test_tools = ["click"]
+all=["click"]
+
+
+[tool.poetry.group.dev.dependencies]
```
[the remark this answers is not in the document]
Comment (solarw) [reply pyproject.toml:62]: done

File: pyproject.toml (hunk 23 above)
[the remark this answers is not in the document]
Comment (solarw) [reply pyproject.toml:110]: it's required by http server connection.
newer versions does not work out of the box
poetry fails to lock with version specified.


need to open an issue to update http server connection package with requirements updated

File: packages/fetchai/skills/tac_control/helpers.py (hunk 7 above)
Comment (solarw) [reply packages/fetchai/skills/tac_control/helpers.py:321]: if i take it out, it will not pass CI at all.

File: tox.ini (hunk 16 above)
Comment (solarw) [reply tox.ini:23]: required by http server connection. does not work with the latest modules. so need to update package code

File: tox.ini (hunk 17 above)
Comment (solarw) [reply tox.ini:28]: typed ast fix required for the old black version.

File: tox.ini (hunk 19 above)
Comment (solarw) [reply tox.ini:180]: click is required to set up a specific version compatible with current mypy

File: tox.ini (hunk 18 above)
Comment (5A11) [reply tox.ini:95]: Going away ini the black PR #2782?

File: tox.ini (hunk 17 above)
Comment (5A11) [reply tox.ini:28]: Going away ini the black PR #2782?

File: tox.ini [hunk 26]
```diff
@@ -141,58 +71,64 @@ commands = {[plugins]commands}
...
 [testenv:black-check]
 skipsdist = True
 skip_install = True
-deps =
-    black==19.10b0
-commands =black aea benchmark examples packages plugins scripts tests --check --verbose
+commands = 
+  poetry run python ./install_packages.py click black
+  # click is required to set up a specific version compatible with current black version 19.10b0
+  poetry run black aea benchmark examples packages plugins scripts tests --check --verbose
```
Comment (5A11) [inline tox.ini:104]: Should be cleaned in PR #2782.

File: tox.ini (hunk 18 above)
Comment (solarw) [reply tox.ini:95]: yes, this planned

File: tox.ini (hunk 17 above)
Comment (solarw) [reply tox.ini:28]: yes, this planned

**Review by jrriehl — APPROVED** [review]

**Review by 5A11 — APPROVED** [review]
lgtm
